toqito/nonlocal_games/bell_inequality_max.py

```diff
@@ -0,0 +1,436 @@
+    :raises ValueError: If the NPA level `k` is invalid.
+    :raises ValueError: If generating NPA constraints fails.
+    :raises cvxpy.error.SolverError: If the CVXPY solver fails.
+
+    """
+    oa, ob, ma, mb = desc
+    mtype_low = mtype.lower()
+    notation_low = notation.lower()
+
+    if notation_low not in ["fp", "fc", "cg"]:
+        raise ValueError(f"Invalid notation: {notation}. Must be 'fp', 'fc', or 'cg'.")
+
+    bmax = None
+    problem = None
+
+    if mtype_low == "nosignal":
+        if notation_low == "fc" and (oa != 2 or ob != 2):
+            raise ValueError(
+                "Notation conversion failed: 'fc' notation is only supported for binary outputs (oa=2, ob=2)."
+            )
+        try:
+            if notation_low == "cg":
+                M = coefficients
+            elif notation_low == "fp":
+                M = fp_to_cg(coefficients, behaviour=False)
+            else:
+                M = fc_to_cg(coefficients, behaviour=False)
+        except ValueError as e:
+            raise ValueError(f"Notation conversion failed: {e}") from e
+
+        cg_dim = ((oa - 1) * ma + 1, (ob - 1) * mb + 1)
+        if M.shape != cg_dim:
+            raise ValueError(
+                f"Coefficient shape {M.shape} incompatible with desc={desc} and CG notation (expected {cg_dim})."
+            )
+
+        p_cg = cvxpy.Variable(cg_dim, name="p_cg")
+
+        objective = cvxpy.Maximize(cvxpy.sum(cvxpy.multiply(M, p_cg)))
+
+        constraints = [p_cg[0, 0] == 1]
+        fp_expressions = _cg_to_fp_cvxpy(p_cg, desc)
+
+        constraints += [expr >= -tol for expr in fp_expressions]
+
+        problem = cvxpy.Problem(objective, constraints)
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=UserWarning)
+            bmax = problem.solve(solver=cvxpy.SCS, eps=tol, verbose=False)
+
+        if problem.status not in [cvxpy.OPTIMAL, cvxpy.OPTIMAL_INACCURATE]:
+            print(f"Warning: Solver status for 'nosignal': {problem.status}")
+            if problem.status in [cvxpy.INFEASIBLE, cvxpy.INFEASIBLE_INACCURATE]:
+                bmax = -np.inf
+            elif problem.status in [cvxpy.UNBOUNDED, cvxpy.UNBOUNDED_INACCURATE]:
+                bmax = np.inf
+
+    elif mtype_low == "quantum":
+        if notation_low == "fc" and (oa != 2 or ob != 2):
+            raise ValueError(
+                "Notation conversion failed: 'fc' notation is only supported for binary outputs (oa=2, ob=2)."
+            )
+
+        if not isinstance(k, (str, numbers.Integral)) or (isinstance(k, numbers.Integral) and k < 0):
+            raise ValueError(f"Invalid NPA level k={k}. Must be a non-negative integer or a valid string level.")
+        if isinstance(k, str):
+            valid_k_prefixes = ["1", "2", "3", "4"]
+            valid_k_suffixes = ["", "+a", "+b", "+ab", "+ba", "+aab", "+abb", "+aba", "+bab"]
```

Comment:
I know computationally that it's unlikely the user will be able to compute beyond level `4`, but it's possible. We may want to not hardcode these as strings, and instead, use some kind of regex, which would simplify the checks you have below that loop through the string. Something like the following:

```python
bool(re.fullmatch(r"\d+(\+[ab]+)?", s))
```